KAFKA-17607: Add CI step to verify LICENSE-binary

xijiu opened the pull request:
Add the verify_license.py script to our build to detect missing licenses. 

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Ken Huang <s7133700@gmail.com>, David Arthur <mumrah@gmail.com>

xijiu committed:
finish coding

xijiu commented:
I added a new gradlew task `libLicenseCheck`, so when execute command  `./gradlew clean libLicenseCheck`, it will print the missing libs in license file

![image](https://github.com/user-attachments/assets/096cf2c7-3881-4a2f-88d6-83b2b17e12e8)

@mumrah @chia7712  @chiacyu    PTAL  ☺

m1a2st reviewed (commented):
Hello @xijiu, Thanks for this PR.
Pardon me, I have a question that this `libLicenseCheck` gradle task will be execute on our CI, should we add verify step on the `build.yml`?

xijiu committed:
add libLicenseCheck to CI

xijiu commented:
> Hello @xijiu, Thanks for this PR. Pardon me, I have a question that this `libLicenseCheck` gradle task will be exected on our CI, should we add verify step on the `build.yml`?

@m1a2st  Thanks for the CR. I have added this to the CI, PTAL

mumrah commented on .github/workflows/build.yml:
```diff
@@ -143,6 +143,9 @@ jobs:
           fi
+      - name: Check missing lib license
+        run: |
+          ./gradlew libLicenseCheck
```
Can you add this to the previous `gradlew` invocation?

mumrah commented on build.gradle:
```diff
@@ -1328,6 +1328,43 @@ project(':core') {
     duplicatesStrategy 'exclude'
   }
 
+  task libLicenseCheck(dependsOn: 'releaseTarGz', type: Copy) {
```
I think it would be better to do this as part of releaseTarGz so we can't mistakenly produce an artifact that is missing licenses. I glanced through build.gradle and I don't see where we're gathering the libraries into the LICENSE file. Do you know where that happens?

xijiu replied:
hi David, do you mean where we replace the content of LICENSE with the content of LICENSE-binary ? 
If that's the case, I think it should be below:
`from "$rootDir/NOTICE-binary" rename {String filename -> filename.replace("-binary", "")}`  in the releaseTarGz task

mumrah reviewed (commented):
Thanks for the patch @xijiu!

xijiu committed:
move lib license to task releaseTarGz

xijiu committed:
revert

m1a2st commented on .github/workflows/build.yml:
```diff
@@ -143,6 +143,9 @@ jobs:
             find ./site-docs/generated -type f -exec grep -L "." {} \; >&2
             exit 1
           fi
+      - name: Check missing lib license
+        run: |
+          ./gradlew releaseTarGz
```
I think @mumrah means, this gradle command should move to `Compile and validate` step

xijiu committed:
add releaseTarGz to releaseTarGz

xijiu commented:
@mumrah @m1a2st    I have fixed it, PTAL

m1a2st commented on .github/workflows/build.yml:
```diff
@@ -114,6 +114,7 @@ jobs:
         # --no-scan:      For public fork PRs, we won't attempt to publish the scan
         run: |
           ./gradlew --build-cache --info $SCAN_ARG check siteDocTar -x test
+          ./gradlew releaseTarGz
```
how about `./gradlew --build-cache --info $SCAN_ARG check siteDocTar releaseTarGz -x test`

xijiu replied:
Nice idea, I will update it.

xijiu committed:
fix code

mumrah commented on .github/workflows/build.yml:
```diff
@@ -113,7 +113,7 @@ jobs:
         # --no-scan:      For public fork PRs, we won't attempt to publish the scan
         run: |
-          ./gradlew --build-cache --info $SCAN_ARG check siteDocTar -x test
+          ./gradlew --build-cache --info $SCAN_ARG check siteDocTar releaseTarGz -x test
```
releaseTarGz includes siteDocTar, so I think we can exclude siteDocTar here

xijiu replied:
hi @mumrah , I have excluded `siteDocTar`

chia7712 commented on build.gradle:
```diff
@@ -1326,6 +1326,42 @@ project(':core') {
+            }
+          }
+          if (!isLicensePresent) {
+            println "warning: ${file.name} is missing in the license file"
```
Can we find out the license which is in the license file but is not in distribution?

xijiu replied:
@chia7712 
It seems quite necessary, but the formats of these jar licenses are not quite the same, such as:
```
jakarta.annotation-api-1.3.5
zstd-jni-1.5.6-6 see: licenses/zstd-jni-BSD-2-clause
argparse4j-0.7.0, see: licenses/argparse-MIT
```

We should extract the following strings
```
jakarta.annotation-api-1.3.5
zstd-jni-1.5.6-6
argparse4j-0.7.0
```

So if we want to extract those, the rules could be: 

1. No spaces in each line 
2. Contains spaces, but includes ' see:' 
3. Contains spaces, but includes ', see:'

But as the subsequent modifications are made, this may be take mistakes.

Or we can establish a guideline to identify lib-licenses (such as add specified prefix or suffix), which should be followed when modifying `LICENSE-binary` in the future.

WDYT ?

chia7712 replied:
@xijiu Could you please open a jira to trace this? I'd like to merge this PR first.

xijiu replied:
@chia7712    
Sure, file:  https://issues.apache.org/jira/browse/KAFKA-18420     
And I will continue to work on the issue.

chia7712 reviewed (commented):
@xijiu thanks for this great added CI task.

chia7712 commented:
@xijiu btw, I open https://issues.apache.org/jira/browse/KAFKA-18375 to fix the license file for 4.0 release.

xijiu committed:
exclude siteDocTar

chia7712 commented:
@xijiu Could you please test this PR again? I assume there is no error as [KAFKA-18375](https://issues.apache.org/jira/browse/KAFKA-18375) had been fixed.

xijiu commented:
@chia7712 
I ran the `./gradlew clean releaseTarGz` again after I merged the `upstream/trunk` branch. And the test was successful as below:

<img width="393" alt="image" src="https://github.com/user-attachments/assets/12edb2d2-ecc1-4570-9dc5-dbe04d38db50" />

chia7712 commented on build.gradle:
```diff
@@ -1326,6 +1326,42 @@ project(':core') {
+      if (result)
+        println "LIBS LICENSE check succeed !"
+      else
+        println "LIBS LICENSE check failed !"
```
should we mark the task failed to ensure developers have to fix it?

xijiu committed:
mark the task failed to ensure developers have to fix it

xijiu commented:
@chia7712   I have added a `GradleException` to fail the task when there are missing lib licenses.  
PTAL

fonsdant commented:
Hi, @mumrah! I have noted this task is very similiar to another old one I had picked up. This is my PR: #17888. Are they the same?

https://issues.apache.org/jira/browse/KAFKA-12622

And it seems to me related to @chia7712's comment too:

> Can we find out the license which is in the license file but is not in distribution?

xijiu commented:
@chia7712 @mumrah   PTAL  🤝

chia7712 commented:
@xijiu #18931 offers a script for license verification. could you please consider integrating it into CI?

xijiu committed:
for github action test

xijiu committed:
revert

xijiu commented:
@chia7712  I have added it, PTAL.

For test, I added the invalid content to `LICENSE-binary`:
<img width="490" alt="image" src="https://github.com/user-attachments/assets/eb505eb8-5284-449c-b6c3-d00b19255891" />

Then triggered CI and failed:
<img width="966" alt="image" src="https://github.com/user-attachments/assets/9a21df97-c2fd-4664-9410-4b420988e262" />
<img width="1161" alt="image" src="https://github.com/user-attachments/assets/542ff715-a3e6-429f-bb77-5695bb0580a5" />

So the verify license file has taken effect.

m1a2st commented on .github/workflows/build.yml:
```diff
@@ -127,6 +127,9 @@ jobs:
           gradle-cache-write-only: ${{ inputs.is-trunk }}
           develocity-access-key: ${{ secrets.DEVELOCITY_ACCESS_KEY }}
+      - name: Verify license file
+        if: always()
```
`always()` means this will run even if a previous step failed. I think we don't need this in this step, WDYT?

xijiu replied:
@m1a2st  Thanks for the code review, and you are right, I have fixed it, PTAL

xijiu committed:
bug fix

chia7712 commented on .github/workflows/build.yml:
```diff
@@ -127,6 +127,8 @@ jobs:
           gradle-cache-write-only: ${{ inputs.is-trunk }}
           develocity-access-key: ${{ secrets.DEVELOCITY_ACCESS_KEY }}
+      - name: Verify license file
+        run: python committer-tools/verify_license.py
```
the script needs to run the `clean releaseTarGz`, so it will rebuild whole project and it seems to me it is a bit expensive as CI. Maybe we can do more works on this?

1. run this check after `Check generated documentation`
2. `Compile and validate` should build `releaseTarGz` too
3. `verify_license.py` should be able to skip  `releaseTarGz`

xijiu committed:
skip releaseTarGz

xijiu committed:
exclude docs key

xijiu committed:
add releaseTarGz to

xijiu commented:
@chia7712  I have added a ArgumentParser to `verify_license.py` and it cloud be able skip the `releaseTarGz` now:

Skip:
```
python verify_license.py --skipReleaseTarGz true
```

NOT Skip:
```
python verify_license.py
or
python verify_license.py --skipReleaseTarGz false
```

PTAL  🤝

xijiu committed:
remove siteDocTar

chia7712 commented on committer-tools/verify_license.py:
```diff
@@ -74,12 +75,20 @@ def get_license_deps(license_text):
+        print("Skip running './gradlew clean releaseTarGz'")
+    else:
+        run_gradlew(project_dir)
     # Build the tarball.
```
please move the comment to line#89

chia7712 commented on committer-tools/verify_license.py:
```diff
@@ -74,12 +75,20 @@ def get_license_deps(license_text):
 
 def main():
+    # Argument parser
+    parser = argparse.ArgumentParser(description="Process some integers.")
```
please replace `Process some integers.` by other description

chia7712 commented on committer-tools/verify_license.py:
```diff
@@ -74,12 +75,20 @@ def get_license_deps(license_text):
 def main():
+    # Argument parser
+    parser = argparse.ArgumentParser(description="Process some integers.")
+    parser.add_argument('--skipReleaseTarGz', type=str, default='false', help='Whether to skip ReleaseTarGz (true/false)')
```
`default='false'` -> `default=False`

chia7712 commented on committer-tools/verify_license.py:
```diff
@@ -74,12 +75,20 @@ def get_license_deps(license_text):
     print("Using project directory:", project_dir)
-    
+
+    if args.skipReleaseTarGz.lower() == 'true':
```
`if args.skipReleaseTarGz:`

chia7712 commented on .github/workflows/build.yml:
```diff
@@ -166,6 +166,8 @@ jobs:
             exit 1
           fi
+      - name: Verify license file
+        run: python committer-tools/verify_license.py --skipReleaseTarGz true
```
`true` is redundant.

chia7712 reviewed (commented):
@xijiu thanks for this patch.

xijiu committed:
adjust code

xijiu commented:
@chia7712   Thanks for CR and I have fixed them.  There is a small point that when I executed the python script locally, it prompted that I needed to provide parameters.  Please take a look.

mumrah commented on committer-tools/verify_license.py:
```diff
@@ -74,12 +75,20 @@ def get_license_deps(license_text):
 def main():
+    # Argument parser
+    parser = argparse.ArgumentParser(description="Whether to skip executing ReleaseTarGz.")
+    parser.add_argument('--skipReleaseTarGz', type=str, default=False, help='Whether to skip ReleaseTarGz (true/false)')
```
I would suggest a flag name like "rebuild" or similar. 

Also, this is not the correct way to configure a boolean argument. You should do something like:

```python
parser.add_argument("--my-flag", action="store_true", help="Help text")
```

https://docs.python.org/3/library/argparse.html#action

chia7712 replied:
> I would suggest a flag name like "rebuild" or similar.

The script should build the distribution if the flag is unset, as it shouldn't assume a distribution already exists. Perhaps the flag could be named skipBuild?

mumrah replied:
Yea that makes sense. I wasn't sure what default behavior we would want. 

In that case, I'd suggest `--skip-build` :) (i like kebab-case for CLI arguments)

xijiu replied:
@mumrah @chia7712   Thanks for the wonderful suggestions. I have fixed them, PTAL

xijiu committed:
fix

mumrah commented on committer-tools/verify_license.py:
```diff
@@ -77,14 +77,14 @@ def get_license_deps(license_text):
     # Argument parser
     parser = argparse.ArgumentParser(description="Whether to skip executing ReleaseTarGz.")
-    parser.add_argument('--skipReleaseTarGz', type=str, default=False, help='Whether to skip ReleaseTarGz (true/false)')
+    parser.add_argument('--skip-build', action="store_true", help='Whether to skip ReleaseTarGz (true/false)')
```
nit: mixture of single and double quotes. I'd suggest using double quotes like the rest of the file.

Also the `(true/false)` part is not necessary. The help text should say what setting this flag does, e.g. "skip the build"

xijiu replied:
yeah, the details are very important, I have fixed it, PTAL

mumrah reviewed (commented):
One small comment

xijiu committed:
fix

mumrah reviewed (approved):
LGTM, thanks @xijiu!
